src/event/EventManager.ts

```diff
@@ -0,0 +1,54 @@
+/*
+ * Copyright 2020-2023 Delft University of Technology and SynTest contributors
+ *
+ * This file is part of SynTest Framework - SynTest Core.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { Encoding } from "..";
+import { PluginInterface } from "./PluginInterface";
+import { ProgramState } from "./ProgramState";
+
+export class EventManager<T extends Encoding> {
+  private _state: ProgramState<T>;
+  private listeners: PluginInterface<T>[] = [];
+
+  constructor(state: ProgramState<T>) {
+    this._state = state;
+  }
+
+  registerListener(listener: PluginInterface<T>) {
+    this.listeners.push(listener);
+  }
+
+  removeListener(listener: PluginInterface<T>) {
+    this.listeners.splice(this.listeners.indexOf(listener));
+  }
+
+  emitEvent(event: keyof PluginInterface<T>) {
```

Comment:
We could also provide the program state as a parameter of this function

Reply:
We could, but it is not necessary and would require passing the program state to each part of the code which does not seem like a good idea.